addons/mail/static/src/core/common/store_service.js

```diff
@@ -93,19 +73,116 @@ export class Store extends BaseStore {
     odoobotOnboarding;
     users = {};
     internalUserGroupId = null;
-    registeredImStatusPartners = null;
+    imStatusTrackedPersonas = Record.many("Persona", {
+        compute() {
+            return Object.values(this.Persona?.records ?? []).filter(
+                (persona) =>
+                    persona.type === "partner" &&
+                    persona.im_status !== "im_partner" &&
+                    !persona.is_public
+            );
+        },
+        onUpdate() {
+            this.updateImStatusRegistration();
+        },
+        eager: true,
+    });
     hasLinkPreviewFeature = true;
     // messaging menu
     menu = { counter: 0 };
+    menuThreads = Record.many("Thread", {
+        /** @this {import("models").Store} */
+        compute() {
+            /** @type {import("models").Thread[]} */
+            let threads = Object.values(this.Thread.records).filter(
+                (thread) =>
+                    thread.displayToSelf ||
+                    (thread.needactionMessages.length > 0 && thread.type !== "mailbox")
+            );
+            const tab = this.discuss.activeTab;
+            if (tab !== "main") {
+                threads = threads.filter(({ type }) => this.tabToThreadType(tab).includes(type));
+            } else if (tab === "main" && this.env.inDiscussApp) {
+                threads = threads.filter(({ type }) =>
+                    this.tabToThreadType("mailbox").includes(type)
+                );
+            }
+            return threads;
+        },
+        /**
+         * @this {import("models").Store}
+         * @param {import("models").Thread} a
+         * @param {import("models").Thread} b
+         */
+        sort(a, b) {
+            /**
+             * Ordering:
+             * - threads with needaction
+             * - unread channels
+             * - read channels
+             * - odoobot chat
+             *
+             * In each group, thread with most recent message comes first
+             */
+            if (a.correspondent?.eq(this.odoobot) && !b.correspondent?.eq(this.odoobot)) {
+                return 1;
+            }
+            if (b.correspondent?.eq(this.odoobot) && !a.correspondent?.eq(this.odoobot)) {
+                return -1;
+            }
+            if (a.needactionMessages.length > 0 && b.needactionMessages.length === 0) {
+                return -1;
+            }
+            if (b.needactionMessages.length > 0 && a.needactionMessages.length === 0) {
+                return 1;
+            }
+            if (a.message_unread_counter > 0 && b.message_unread_counter === 0) {
+                return -1;
+            }
+            if (b.message_unread_counter > 0 && a.message_unread_counter === 0) {
+                return 1;
+            }
+            if (!a.newestPersistentMessage?.datetime && b.newestPersistentMessage?.datetime) {
```

Comment:
unrelated to PR, but for channels we should use last_interest_dt in priority before checking last message date (order should be the same as in discuss sidebar for channels)